WordPress/gutenberg
# Improve component CSS reliability with scoped prefix

This update adds a new (experimental) component in `@wordpress/components` that improves the integration with `wp-admin` for generated CSS-in-JS styles.

### StyledScopeProvider

StyledScopeProvider is a component that adds CSS selector prefixing for styles rendered with [Emotion](https://github.com/emotion-js/emotion) (currently experimental within `@wordpress/components`).

Prefixing increases style specificity, allowing for a smoother integration into environments with pre-existing CSS rules.

<img width="557" alt="Screen Shot 2020-06-22 at 11 53 30 AM" src="https://user-images.githubusercontent.com/2322354/85308293-0a9b0800-b47f-11ea-924a-fbf989b0efbd.png">


(Example: Generated scoped styles for `<NumberControl />` within Gutenberg.

With the new `StyledScopeProvider`, we can remove the (previous) `&&&` work-arounds. These were added to manually bump up specificity for overriding `wp-admin`/Editor styles.

Note: At this moment, there are only a handful of Emotion styled `@wordpress/components` Components. 

### Extra History

This was a technique I developed for a [previous project in 2018](https://github.com/emotion-js/emotion/issues/760#issuecomment-404353706). It was used to reliably integrate and render a React components into an old PHP/jQuery powered application (much like WordPress!).

The code I'm submitted was abstracted from a library I created:
https://github.com/ItsJonQ/styled-providers

## Usage

```jsx
import {
	__experimentalStyledScopeProvider as StyledScopeProvider,
	__experimentalInputControl as InputControl,
} from '@wordpress/components';

const Example = () => (
	<StyledScopeProvider scope="html body.wp-admin">
		...
		<InputControl />
		...
	</StyledScopeProvider>
);
```

In the above example, the `<InputControl />` component styles (which are generated with [Emotion](https://github.com/emotion-js/emotion)), will be prefixed with `html body.wp-admin`.
Instead of the generated CSS selector of `.css-123jda`, it would instead be `html body.wp-admin .css-123jda`. Increasing specificity in a controlled manner is helpful in situations where there may be conflicting styles.

In the case of `<InputControl />`, existing style rules such as `.block-editor input[type="text]` may pose problematic. However, the automatically generated selectors `html body.wp-admin .css-123jda` will help ensure the elements are styled as expected.


## How has this been tested?

This has been tested in both Storybook and local Gutenberg.

* Run `npm install`
* Run `npm run dev`
* Open up the editor
* Add a Cover block
* Make sure the min-height input is styled correctly and works
* Add a Column block
* Ensure the range slider is styled correctly and works

## Types of changes

* Add `@emotion` dependencies to `@wordpress/components`
* Add new `<StyledScopeProvider />` component
* Export as `__experimental`
* Integrate into Editor Provider
* Integrate into Edit Site Editor

## Checklist:

Resolves: https://github.com/WordPress/gutenberg/issues/18483
